### PR TITLE
Add cause to RealmMigrationNeededException

### PR DESCRIPTION
**Pull request by beeender**

### CODE REVIEW

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 1]
```diff
@@ -1344,7 +1344,23 @@ private void checkMaxDepth(int maxDepth) {
...
+    }
+
+    /**
+     * Called when migration needed in the Realm initialization.
+     *
+     * @param configuration {@link RealmConfiguration}
+     * @param cause which triggers this migration.
+     * @throws FileNotFoundException if the Realm file doesn't exist.
+     */
+    public static void migrateRealm(final RealmConfiguration configuration, final RealmMigrationNeededException cause)
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: This should be package private?

File: CHANGELOG.md [hunk 2]
```diff
@@ -11,6 +11,10 @@
 * Fixed a bug causing the Realm Transformer to not transform field access in the model's constructors (#3361).
 * Fixed a bug causing a build failure when the Realm Transformer adds accessors to a model class that was already transformed in other project (#3469).
 
+### Enhancements
+
+* A `RealmMigrationNeededException` will be thrown with a cause to show the detailed message when migration is needed and the migration block is not in the `RealmConfiguration`.
```
Comment (cmelchior) [inline CHANGELOG.md]: when a migration is needed

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 3]
```diff
@@ -1344,7 +1344,23 @@ private void checkMaxDepth(int maxDepth) {
      * @throws FileNotFoundException if the Realm file doesn't exist.
      */
     public static void migrateRealm(RealmConfiguration configuration) throws FileNotFoundException {
-        migrateRealm(configuration, null);
+        migrateRealm(configuration, (RealmMigration)null);
```
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: need a space before `null`

File: realm/realm-library/src/main/java/io/realm/Realm.java (hunk 1 above)
Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: oops.. copy & paste error

File: realm/realm-library/src/main/java/io/realm/BaseRealm.java [hunk 4]
```diff
@@ -706,15 +706,17 @@ static boolean compactRealm(final RealmConfiguration configuration) {
...
     protected static void migrateRealm(final RealmConfiguration configuration, final RealmMigration migration,
-                                       final MigrationCallback callback) throws FileNotFoundException {
+                                       final MigrationCallback callback, final RealmMigrationNeededException cause)
+            throws FileNotFoundException {
         if (configuration == null) {
             throw new IllegalArgumentException("RealmConfiguration must be provided");
         }
         if (migration == null && configuration.getMigration() == null) {
-            throw new RealmMigrationNeededException(configuration.getPath(), "RealmMigration must be provided");
+            throw new RealmMigrationNeededException(configuration.getPath(), "RealmMigration must be provided", cause);
```
Comment (kneth) [inline realm/realm-library/src/main/java/io/realm/BaseRealm.java]: So an exception is taking an exception as argument?

**Comment by zaki50** [comment]
:+1: other than my and cm's comments

File: realm/realm-library/src/main/java/io/realm/BaseRealm.java (hunk 4 above)
Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/BaseRealm.java]: @kneth See this https://realm.io/docs/java/1.2.0/api/io/realm/exceptions/RealmMigrationNeededException.html#RealmMigrationNeededException-java.lang.String-java.lang.String-java.lang.Throwable- :tongue:

File: realm/realm-library/src/main/java/io/realm/Realm.java (hunk 1 above)
Comment (Zhuinden) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: Private or package private?

Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: `private`~~

Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: Probably `private` anything but `public` at least 😄

**Pull request merged**